### ISSUE TITLE
Problem with vi-mode after changing $EDITOR variable

### ISSUE THREAD

**Issue by longshi0** [robbyrussell/oh-my-zsh#5210]
Happy oh-my-zsh user and recent vim-to-spacemacs convert here.

After I replaced my $EDITOR variable from "vim" to "emacsclient", when i try to invoke edit-command-line with 'v', i get this: 
_ERROR_: Symbol's value as variable is void: /tmp/zshEjbopL

Is it possible to have 'emacsclient' set as $EDITOR and still be able to edit commands within editor?

**Comment by pawsen** [comment]
Hi Michael,

I have the same problem. In .oh-my-zsh/plugins/emacs/emacs.plugin.zsh:
```
 export EMACS_PLUGIN_LAUNCHER="$ZSH/plugins/emacs/emacsclient.sh"
 export EDITOR="${EDITOR:-${EMACS_PLUGIN_LAUNCHER}}
```
Just to be sure it is not the configuration of Emacs that is causing trouble, we start a new emacs daemon without loading any init-files
```
\emacs -q --daemon
```
(the \ is to avoid using the zsh-alias) and invoke edit-command-line (also as C-x C-e) I get the same error.

Setting  export EDITOR='/usr/bin/emacs  -nw'  or export EDITOR='/usr/bin/emacs'  works as expected.

If I start an bash (instead of zsh) session, then export EDITOR='/usr/bin/emacsclient' works as expected.

If I start a zsh session without loading oh-my-zsh, eg. mv ~/.zshrc{,.bak} and run the relevant commands:
```
autoload -U edit-command-line
zle -N edit-command-lin
bindkey '\C-x\C-e' edit-command-line
EDITOR='/usr/bin/emacsclient' 
```
I get the same error. So the problem seems to be with zsh, not emacs or oh-my-zsh.

**Comment by jrosdahl** [comment]
I have done some research on this annoying behavior and I found the culprit: it's a bug in zsh 5.1 and 5.1.1. Fixed in 5.2.

See http://www.zsh.org/mla/workers//2015/msg02489.html for the bug report. The fix in zsh's Git repo looks like this:

```
commit 56ed4df8988249338b0f8f2d999df4217ca533e6
Author: Barton E. Schaefer <schaefer@zsh.org>
Date:   Tue Sep 22 08:52:52 2015 -0700

    36587: use +LINE:COLUMN to place the cursor when invoking emacs variants, for emacsclient

diff --git a/ChangeLog b/ChangeLog
index e5ea4fd..6d69e6d 100644
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,3 +1,8 @@
+2015-09-22  Barton E. Schaefer  <schaefer@zsh.org>
+
+       * 36587: Functions/Zle/edit-command-line: use +LINE:COLUMN to
+       place the cursor when invoking emacs variants, for emacsclient
+
 2015-09-22  Peter Stephenson  <p.stephenson@samsung.com>
 
        * 36586: Han Pingtian: Src/Zle/compmatch.c: tweak to completion
diff --git a/Functions/Zle/edit-command-line b/Functions/Zle/edit-command-line
index 2c7f34b..103a1c1 100644
--- a/Functions/Zle/edit-command-line
+++ b/Functions/Zle/edit-command-line
@@ -11,13 +11,16 @@
 
   # Compute the cursor's position in bytes, not characters.
   setopt localoptions nomultibyte
-  integer byteoffset=$(( $#PREBUFFER + $#LBUFFER + 1 ))
 
   # Open the editor, placing the cursor at the right place if we know how.
   local editor=${${VISUAL:-${EDITOR:-vi}}}
   case $editor in 
-    (*vim*) ${=editor} -c "normal! ${byteoffset}go" -- $1;;
-    (*emacs*) ${=editor} $1 -eval "(goto-char ${byteoffset})";;
+    (*vim*)
+      integer byteoffset=$(( $#PREBUFFER + $#LBUFFER + 1 ))
+      ${=editor} -c "normal! ${byteoffset}go" -- $1;;
+    (*emacs*)
+      local lines=( ${(f):-"$PREBUFFER$LBUFFER"} )
+      ${=editor} +${#lines}:$((${#lines[-1]} + 1)) $1;;
     (*) ${=editor} $1;;
   esac
 
```

Here's the workaround I use: Make a wrapper script like

```
#!/bin/sh

exec emacsclient "$@"
```

 with a name not matching `*emacs*` (nor `*vim*` :)) and set `EDITOR` to it.

**Comment by mcornella** [comment]
Thanks for the explanation!
